### PR TITLE
feat(distributed): enforce a thinking budget on the rank's decode path

### LINKED ISSUE #2711: thinking_budget is rejected by distributed inference (and would be bypassed anyway), so clusters get unbounded xhigh reasoning or nothing
**Is your feature request related to a problem? Please describe.**
On a distributed TP2 cluster (oMLX 0.6.0.dev1, 2× M4 Max 48 GB, ring), Qwen3.8-27B's default reasoning is xhigh and there's no way to bound it: `thinking_budget` is rejected by the distributed engine, and even when forwarded the budget is not enforced. Agentic sessions either run unbounded xhigh (burning the whole max_tokens budget in reasoning, cut off at finish_reason=length before a tool call) or run with reasoning fully off. There's no middle ground.

The distributed bypass is separate from the single-node speculative-decode gap (#2455, which is about the vlm_mtp path): the distributed ranks run `mlx_lm.server`, which builds logits processors once from CLI args and never reads a per-request `thinking_budget` from the chat body. So even though the coordinator can forward `thinking_budget` to rank zero, nothing on the rank enforces it.

**Describe the solution you'd like**
Support a thinking budget (or reasoning_effort) in distributed decode, so a cluster can cap reasoning at e.g. 2048-4096 tokens and still get its answer/tool call out. For the distributed path this means patching the rank's `mlx_lm.server` request handling to read `thinking_budget` per request and inject oMLX's `ThinkingBudgetProcessor` (omlx/api/thinking.py) into the logits processors it builds at decode time.

**Describe alternatives you've considered**
- Template-level `enable_thinking` on/off only (works today, but all-or-nothing).
- Raising the output budget so the model has room to reason and act (current workaround, 32k).
- Per-prompt nudges to stop reasoning (unreliable — the model keeps analyzing).
- `reasoning_effort` (low/medium/xhigh) as a template instruction: works through distributed today via `chat_template_kwargs` (no decode-loop involvement), so it is the practical near-term workaround, but it is advisory — it does not hard-cap thinking tokens.

**Additional context**
I verified template-level `chat_template_kwargs.enable_thinking` does reach the ranks on distributed (an `enable_thinking:false` request came back with zero reasoning), so the forwarding path exists — it's the budget enforcement that's missing. `reasoning_effort` similarly reaches the ranks (the template renders it), so per-request effort is enforceable today; only a hard token budget needs the rank-side decode patch. Related (single-node speculative path only, not distributed): #2455.

### PR DESCRIPTION
**Pull request by xunlinkx**
On a TP2 cluster, `thinking_budget` was rejected outright. That's understandable: the rank runs mlx-lm's server, which builds its logits processors once from CLI args and never looks at the request body, so forwarding a budget would have been silently ignored — worse than an error.

This makes it actually work. Two pieces:

- The coordinator folds `thinking_budget` into the `chat_template_kwargs` it sends to rank zero (the field mlx-lm already parses per request), instead of dropping it.
- The inference worker wraps mlx-lm's `_make_logits_processors` so a request carrying a budget gets omlx's `ThinkingBudgetProcessor` appended — the same processor single-machine inference uses. The rank resolves think tokens from its own tokenizer, with the same fallbacks the scheduler uses. Models without thinking support are untouched (the wrap is a no-op).

Tested live on a TP2 ring (2x M4 Max 48GB, Qwen3.8-27B): a request with `thinking_budget=16` cut reasoning to ~12 tokens and still produced a complete answer; `thinking_budget=4` combined with `reasoning_effort=low` went straight to the result.

Tests cover the token resolution fallbacks, the per-request injection, and the payload folding. 110 tests pass across the three touched suites.

Fixes #2711

### CODE REVIEW

**Comment by xunlinkx** [comment]
@copilot-pull-request-reviewer please review

File: omlx/cluster/inference_worker.py [hunk 1]
```diff
@@ -87,6 +87,85 @@ def generate_on_rank_stream(instance: Any) -> Any:
...
+                think_end_ids = list(ids)
+        except Exception:
+            pass
+    if think_end_ids is None:
+        try:
+            tid = tokenizer.convert_tokens_to_ids("</think>")
+            if tid != getattr(tokenizer, "unk_token_id", None):
+                think_end_ids = [tid]
+        except (AttributeError, KeyError, TypeError):
+            pass
```
[the remark this answers is not in the document]
Comment (xunlinkx) [reply omlx/cluster/inference_worker.py:129]: The encode call has retry logic: it tries add_special_tokens=False first, catches TypeError and retries without the kwarg, and catches any other Exception. The fallback chain also tries tokenizer.think_end (the attribute), then convert_tokens_to_ids, so it doesn't rely solely on the hard-coded string. Added tests for each fallback path.

File: tests/test_distributed_engine.py [hunk 2]
```diff
@@ -143,7 +165,67 @@ async def test_distributed_stream_bounds_rank_zero_read_stalls():
     finally:
         await engine._client.aclose()
```
[the remark this answers is not in the document]
Comment (xunlinkx) [reply tests/test_distributed_engine.py:145]: The assertion is still there — test_distributed_stream_bounds_rank_zero_read_stalls at line 146 asserts len(status_calls) == 2. The preflight test was changed because thinking_budget is now supported and no longer rejected, but the stream timeout test still verifies availability is rechecked.

**Comment by jundot** [comment]
Thanks for wiring the per-request thinking budget into the distributed rank decode path and for the TP2 validation. I merged this and followed up by allowing the model setting, applying the processor only when reasoning is active, and matching the single-node close-pattern and UTF-8 boundary handling.
